File: dijkstra.hs
-- Author : Furkan BEKAR
-- Date : May 2020
--İNFO : Bu algoritma ağaçlarda en küçük maliyeti hesaplamamıza yarayan algoritmadır.
--       Genellikle Google Maps gibi uygulamalarda gideceğimiz yere ulaşmak için en kısa yolu
--       tarif etmek gibi amaçlar için kullanılır.
-- ################ Dijkstra Short Path Algorithm ################

module Graphs where
--Burada gerekli olan kütüphanelerimizi import ediyoruz ve gerekli olan değişkenlerimizi tanımlıyoruz.
import qualified Data.Set as Set
import qualified Data.Map.Strict as Map

type Mesafeler a = Map.Map (Kose a) Int
type Gidilen a = Map.Map (Kose a) (Kose a)

newtype Kose a = Keyed a
  deriving (Ord, Eq, Show)

-- Burada ise ağacımızı tanımlıyoruz
class Graph g where
  koseler :: Ord a => g a -> Set.Set (Kose a)

  komsular :: Ord a => g a -> Kose a -> Set.Set (Int, Kose a)

newtype MapGraph a = Weighted { toMap :: Map.Map (Kose a) (Set.Set (Int, Kose a)) }

instance Graph MapGraph where
  koseler (Weighted m) =
    Set.union (Map.keysSet m) $ Set.unions $ map (Set.map snd) $ map snd $ Map.toList m

  komsular (Weighted m) v =
    Map.findWithDefault Set.empty v m

-- Bu kısımda artık asıl algoritmayı kurduğumuz kısım ve en kısa yolu bulmada kullandığımız yardımcı fonksiyonu tanımladığımız kısım.
dijkstra :: (Graph g, Ord a) => g a -> Kose a -> (Mesafeler a, Gidilen a)
dijkstra graph source =
  let
    q = koseler graph
    dist = Map.singleton source 0
  in
    aux q dist Map.empty

  where
    aux q dist prev =
      if Set.null q
      then (dist, prev)
      else
        let
          u' = foldr (minIn dist) Nothing q
        in
          case u' of
            Nothing -> (dist, prev)
            Just u ->
              let
                q' = Set.delete u q
                ns = komsular graph u
                (dist', prev') = foldr (insertBetter u) (dist, prev) ns
              in
                aux q' dist' prev'

minIn _ v Nothing = Just v
minIn dist v1 (Just v2) =
  case (Map.lookup v1 dist, Map.lookup v2 dist) of
    (Just d1, Just d2) -> Just (if d1 < d2 then v1 else v2)
    (Nothing, Nothing) -> Nothing
    (Nothing, _) -> Just v2
    (_, Nothing) -> Just v1

insertBetter u (length, v) (dist, prev) =
  case (Map.lookup u dist, Map.lookup v dist) of
    (Just uD, Just vD) ->
      let
        alt = uD + length
      in
        if alt < vD
        then (Map.insert v alt dist, Map.insert v u prev)
        else (dist, prev)
    (Just uD, _) ->
      let
        alt = uD + length
      in
        (Map.insert v alt dist, Map.insert v u prev) 
    _ ->
      (dist, prev)

-- Burda ise algoritmamız son buluyor ve artık konsol ekranına bulduğumuz en kısa yolu yazdırıyoruz.
graph = Weighted $
  Map.fromList
      [
        (Keyed 'H', Set.fromList [(10, Keyed 'W'), (9, Keyed 'C')]),
        (Keyed 'C', Set.fromList [(9, Keyed 'H'), (9, Keyed 'W')]),
        (Keyed '@', Set.fromList [(10, Keyed 'H'), (2, Keyed 'C')])
      ]

main = print $ snd $ dijkstra graph (Keyed 'H')

File: dijkstra.rb
# Author : Furkan BEKAR
# Date : May 2020
#İNFO : Bu algoritma ağaçlarda en küçük maliyeti hesaplamamıza yarayan algoritmadır.
#       Genellikle Google Maps gibi uygulamalarda gideceğimiz yere ulaşmak için en kısa yolu
#       tarif etmek gibi amaçlar için kullanılır.
# ---------------- Dijkstra Short Path Algorithm ----------------

class Graph
Dugumler = Struct.new(:isim, :komsular, :mesafe, :prev)

# Burada yapıcı metodumuzu tanımlıyoruz ve bize gerekli olan değişkenlere atama yapıyoruz.
def initialize(graph)
@koseler = Hash.new{|h,k| h[k]=Dugumler.new(k,[],Float::INFINITY)}
@edges = {}
graph.each do |(v1, v2, mesafe)|
@koseler[v1].komsular << v2
@koseler[v2].komsular << v1
@edges[[v1, v2]] = @edges[[v2, v1]] = mesafe
end
@dijkstra_source = nil
end

# Burada algoritmamızı kurmaya başlıyoruz
def dijkstra(baslangic)
return  if @dijkstra_source == baslangic
q = @koseler.values
q.each do |v|
v.mesafe = Float::INFINITY
v.prev = nil
end
@koseler[baslangic].mesafe = 0
until q.empty?
u = q.min_by {|vertex| vertex.mesafe}
break if u.mesafe == Float::INFINITY
q.delete(u)
u.komsular.each do |v|
vv = @koseler[v]
if q.include?(vv)
alt = u.mesafe + @edges[[u.isim, v]]
if alt < vv.mesafe
vv.mesafe = alt
vv.prev = u.isim
end
end
end
end
@dijkstra_source = baslangic
end

# Burada ise bize en kısa yolu bulmamızı sağlayacak olan yardımcı fonksiyonumuzu tanımlıyoruz.
def shortest_path(baslangic, target)
dijkstra(baslangic)
path = []
u = target
while u
path.unshift(u)
u = @koseler[u].prev
end
return path, @koseler[target].mesafe
end

def to_s
"#<%s koseler=%p edges=%p>" % [self.class.isim, @koseler.values, @edges]
end
end

# Burada ise algoritmamızı TEST ediyoruz
g = Graph.new([ [:ankara, :kayseri, 350],
                [:ankara, :sivas, 450],
                [:ankara, :antalya, 700],
                [:kayseri, :sivas, 500],
                [:kayseri, :igdir, 750],
                [:sivas, :igdir, 550],
                [:sivas, :antalya, 100],
                [:igdir, :van, 300],
                [:van, :antalya, 450],
              ])

baslangic, varis = :ankara, :van
yol, mesafe = g.shortest_path(baslangic, varis)
puts "En kısa yol #{baslangic} dan #{varis} uzaklık ise #{mesafe} km dir:"
puts yol.join(" -> ")

gets.chomp
